tb_plugin/torch_tb_profiler/profiler/op_tree.py

```diff
@@ -0,0 +1,329 @@
+import sys
+from collections import defaultdict
+
+from .. import utils
+from .node import (BackwardNode, ModuleNode, OperatorNode, ProfilerStepNode,
+                   RuntimeNode, is_operator_node)
+from .trace import EventTypes
+
+logger = utils.get_logger()
+
+
+class OpTreeBuilder:
+    BACKWARD_ROOT_PREFIX = 'autograd::engine::evaluate_function:'
+    BACKWARD_ACCUMULATE_GRAD = 'autograd::engine::evaluate_function: torch::autograd::AccumulateGrad'
+
+    def __init__(self, tid2list, tid2zero_rt_list, corrid_to_device, fwd_bwd_map):
+        self.tid2list = tid2list
+        self.tid2zero_rt_list = tid2zero_rt_list
+        self.corrid_to_device = corrid_to_device
+        self.fwd_bwd_map = fwd_bwd_map
+        self.main_tid = None
+        self.tid2tree = None
+
+    def build_tree(self):
+        '''Construct the BackwardNode and replace the original backward nodes
+        '''
+        self.tid2tree = self._build_tree(self.tid2list, self.tid2zero_rt_list, self.corrid_to_device)
+        self._set_main_tid()
+
+        modules, backward_nodes = self._get_modules()
+        if not modules or not backward_nodes:
+            return self.tid2tree
+
+        _, ts2parent = OpTreeBuilder._get_node_parents(backward_nodes)
+        agg_nodes = OpTreeBuilder._group_backward_nodes(backward_nodes)
+        fwd_bwd_root = self._get_backward_roots(self.fwd_bwd_map, ts2parent, agg_nodes)
+        if len(agg_nodes) > 0:
+            logger.warning("some nodes cannot find forward nodes")
+
+        backward_modules = []
+        for module in modules:
+            OpTreeBuilder._build_backward_module(module, None, fwd_bwd_root, backward_modules)
+        OpTreeBuilder._insert_backward_modules(self.tid2tree[self.main_tid], backward_modules)
+        self.tid2tree = {tid: root for tid, root in self.tid2tree.items() if len(root.children) > 0}
+
+        return self.tid2tree
+
+    def _build_tree(self, tid2list, tid2zero_rt_list, corrid_to_device):
+        tid2tree = {}
+
+        staled_device_nodes = []
+        for _, device_nodes in corrid_to_device.items():
+            staled_device_nodes.extend([n for n in device_nodes if n.type == EventTypes.KERNEL])
```

Comment:
this logic maybe should put in the NodeParserMixin.parse_nodes and return the staled_device_nodes from there? Because from outside of the parse_nodes, it is a little strange to understand what 'corrid_to_device' is.

Reply:
good point